includes/subscriptions/class-wc-payments-subscriptions-onboarding-handler.php

```diff
@@ -0,0 +1,127 @@
+<?php
+/**
+ * Class WC_Payments_Subscriptions_Onboarding_Handler
+ *
+ * @package WooCommerce\Payments
+ */
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * A class to handle the onboarding of subscriptions products. Created subscription products will be set to draft until
+ *   onboarding is completed.
+ */
+class WC_Payments_Subscriptions_Onboarding_Handler {
+
+	use WC_Payments_Subscriptions_Utilities;
+
+	/**
+	 * Option for holding an array of product id's to publish post onboarding.
+	 *
+	 * @const string
+	 */
+	const WCPAY_SUBSCRIPTION_AUTO_PUBLISH_PRODUCTS = 'wcpay_subscription_onboarding_products';
+
+	/**
+	 * The account service instance.
+	 *
+	 * @var WC_Payments_Account
+	 */
+	private $account;
+
+	/**
+	 * Constructor
+	 *
+	 * @param WC_Payments_Account $account account service instance.
+	 */
+	public function __construct( WC_Payments_Account $account ) {
+		add_action( 'publish_product', [ $this, 'product_published' ] );
+		add_action( 'woocommerce_payments_account_refreshed', [ $this, 'account_data_refreshed' ] );
+
+		$this->account = $account;
+	}
+
+	/**
+	 * Sets the account service instance reference on the class.
+	 *
+	 * @param WC_Payments_Account $account account service instance.
+	 */
+	public function set_account( WC_Payments_Account $account ) {
+		$this->account = $account;
+	}
+
+	/**
+	 * Convert subscriptions to drafts when using WCPay (without subscriptions) and onboarding is not complete.
+	 *
+	 * @param int $product_id Subscriptions Product id.
+	 */
+	public function product_published( int $product_id ) {
+		// We can skip if the post is not yet marked as published.
+
+		if ( $this->account->is_stripe_connected() ) {
+			return;
+		}
+
+		// If Subscriptions plugin is installed we don't need to do this check.
+		if ( $this->is_subscriptions_plugin_active() ) {
+			return;
+		}
+
+		// Skip products which have already been scheduled or aren't subscriptions.
+		$product = wc_get_product( $product_id );
+		if ( ! $product || ! WC_Subscriptions_Product::is_subscription( $product ) ) {
```

Comment:
For some reason `WC_Subscriptions_Product::is_subscription( $product )` is always returning `false` for me 🤔

Steps to reproduce:

- **Products → Add New**
- Add product name.
- Change product type to "Simple subscription".
- Add price.
- Publish.
- Observe the product is published. (as mentioned above, I debugged this a little and discovered `WC_Subscriptions_Product::is_subscription` returns `false`.

Any ideas?

Reply:
> Any ideas?

It's because of the order of WP hooks. This function is hooked onto the WP core dynamic hook 
`"{$new_status}_{$post->post_type}"` (`'publish_product'`), that runs before the `save_post` hook (which WC hooks into) though and so WC hasn't set the product type yet. 

The default product type is "simple" and so fails this `WC_Subscriptions_Product::is_subscription()` check. 

We will need to use a hook that runs after the product has been saved. eg `woocommerce_admin_process_product_object`